scripts/speech_recognition/oomptimizer.py

```diff
@@ -0,0 +1,525 @@
+                if item["type"] != "dummy"
+            )
+            else "audio"
+        )
+        for direction in ("input", "output")
+    ]
+
+    def get_max_seq_lens(buckets):
+
+        def _determine_lens_for_bucket(bin):
+            if is_2d_bucketing:
+                input_len, output_len = bin
+            else:
+                input_len = bin
+                output_len = ratio * input_len
+            sampling_rate = getattr(
+                model, "sample_rate", 16000
+            )  # TODO: may need to extend schema for broader model coverage
+            match modalities:
+                case "audio", "audio":
+                    return (
+                        compute_num_samples(input_len, sampling_rate=sampling_rate),
+                        compute_num_samples(output_len, sampling_rate=sampling_rate),
+                    )
+                case "audio", "text":
+                    return (compute_num_samples(input_len, sampling_rate=sampling_rate), output_len)
+                case "text", "audio":
+                    return (
+                        input_len,
+                        compute_num_samples(output_len, sampling_rate=sampling_rate),
+                    )
+                case "text", "text":
+                    return input_len, output_len
+                case _:
+                    raise RuntimeError(f"Unexpected modality combination: {_}")
+
+        return [_determine_lens_for_bucket(bin) for bin in buckets]
+
+    click.echo("Starting profiling.")
+    max_seq_lens = get_max_seq_lens(buckets)
+    gen = ProfilingBatchGenerator(schema=schema, start_batch_size=start_batch_size, rel_gap_thresh=threshold)
+    profile = {}
+
+    # Iterate buckets from the largest to the smallest sequences. This usually ends up creating
+    # a tiny bit smaller batches, likely due to worse memory fragmentation.
+    for bucket, (seq_len_in, seq_len_out) in reversed(list(zip(buckets, max_seq_lens))):
+        click.echo(f"The current sequence lengths are: input={seq_len_in} output={seq_len_out}.")
+        gen.reset()
+        batch_idx = 0
+
+        def step():
+            click.echo(
+                f"\t[BEGIN step] [CUDA RAM CURRENT: {torch.cuda.memory_allocated() / (1024 * 1024):.1f}MB] [CUDA RAM MAX: {torch.cuda.max_memory_allocated() / (1024*1024):.1f}MB]"
+            )
+            batch = gen(seq_len_in, seq_len_out)
+            oom = False
+            try:
+                click.echo(f"\tCurrent gap: {gen.current_rel_gap}... ", nl=False)
+                optimizer.zero_grad()
+                out = model.training_step(batch, batch_idx)
+                out['loss'].sum().backward()
+                optimizer.step()
+            except torch.cuda.OutOfMemoryError as e:
+                click.secho(f"OOM!", fg="yellow")
+                oom = True
+            except RuntimeError as e:
+                if "cuFFT error: CUFFT_INTERNAL_ERROR" not in str(e):
```

Comment:
Interesting. Unfortunately there is no way to get the stack of previous cuda errors when we encounter an error like this to ensure that the problem really is an OOM... Yikes.

BTW, I am considering replacing the STFT operation with 1D convolutions as discussed. So I wonder what error cudnn gives if it gets an OOM error, though I'm not sure where it might allocate internally other than for im2col-based GEMM, and even though I think you have to pass the workspace yourself...